python/lsst/daf/butler/_butler.py

```diff
@@ -1673,6 +1673,119 @@ def doImport(importStream: TextIO) -> None:
+            the source butler is using integer IDs. Should not be used
+            if this receiving butler uses integer IDs. Without this dataset
+            import always uses unique.
+
+        Returns
+        -------
+        refs : `list` of `DatasetRef`
+            The refs added to this Butler.
+
+        Notes
+        -----
+        Requires that any dimension definitions are already present in the
+        receiving Butler. The datastore artifact has to exist for a transfer
+        to be made but non-existence is not an error.
+
+        Datasets that already exist in this run will be skipped.
+        """
+        if not self.isWriteable():
+            raise TypeError("Butler is read-only.")
+        progress = Progress("lsst.daf.butler.Butler.transfer_from", level=logging.DEBUG)
+
+        # Will iterate through the refs multiple times so need to convert
+        # to a list if this isn't a collection.
+        if not isinstance(source_refs, collections.abc.Collection):
+            source_refs = list(source_refs)
+
+        log.info("Transferring %d datasets into %s", len(source_refs), str(self))
+
+        if id_gen_map is None:
+            id_gen_map = {}
+
+        # Importing requires that we group the refs by dataset type and run
+        # before doing the import.
+        grouped_refs = defaultdict(list)
+        grouped_indices = defaultdict(list)
+        for i, ref in enumerate(source_refs):
+            grouped_refs[ref.datasetType, ref.run].append(ref)
+            grouped_indices[ref.datasetType, ref.run].append(i)
+
+        # The returned refs should be identical for UUIDs.
+        # For now must also support integers and so need to retain the
+        # newly-created refs from this registry.
+        # Pre-size it so we can assign refs into the correct slots
+        transferred_refs_tmp: List[Optional[DatasetRef]] = [None] * len(source_refs)
+        default_id_gen = DatasetIdGenEnum.UNIQUE
+
+        for (datasetType, run), refs_to_import in progress.iter_item_chunks(grouped_refs.items(),
+                                                                            desc="Importing to registry by "
+                                                                            "run and dataset type"):
+            run_doc = source_butler.registry.getCollectionDocumentation(run)
+            self.registry.registerCollection(run, CollectionType.RUN, doc=run_doc)
+
+            id_generation_mode = default_id_gen
+            if isinstance(refs_to_import[0].id, int):
+                # ID generation mode might need to be overridden when
+                # targetting UUID
+                id_generation_mode = id_gen_map.get(datasetType.name, default_id_gen)
+
+            log.debug("Importing %d refs with %s into run %s",
+                      len(refs_to_import), datasetType.name, run)
+
+            # No way to know if this butler's registry uses UUID.
```

Comment:
Not sure it's a job for this ticket, but I do think we want a public API for this.  It's not just an implementation detail; it's a behavioral change that's important to a lot of higher-level code.

Reply:
It's a transitory problem to a certain extent. Things break if you try to use UUID in an old registry and if I knew that I could strip the ids before calling the import. The question is whether we actually care about that kind of migration. One question I have is whether you are thinking of this as a long-term API or a quick hack. I could easily see a "private" attribute on the dataset manager that we retire when we retire integer IDs. Do you envisage a more public API on `Registry` that will return the `dataset_id_type` even though in the long term that would presumably be the type of `DatasetRef.id`?

Reply:
I think totally public - things in pipe_base and maybe obs_base do care (thinking especially about Quantum and QuantumGraph because I'm working on those now).
I imagine we would probably deprecate this API when we deprecate `int` support, but that depends on how much we want to hedge against changing again in the future.

If we wanted to keep it around forever, we might want to think about whether there should be some other flags, in addition to the type; those might last longer:
- `has_globally_unique_ids`
- `supports_deterministic_ids`

Of course, if we expect both of those to always be true forever after `int` is gone, then maybe just the `int` vs `uuid` type flag is enough.

Reply:
Okay. I'll do this on another ticket because it will also lead to a cleanup of obs_base ingest-raws.